File: packages/sdk-java/src/main/java/com/ubag/sdk/IdempotencyKey.java
package com.ubag.sdk;

import java.security.SecureRandom;

/**
 * Generates ULID-style idempotency keys: 26 characters of Crockford base32,
 * a 10-character millisecond timestamp followed by 16 characters of entropy.
 */
public final class IdempotencyKey {

    private static final char[] CROCKFORD_BASE32 = "0123456789ABCDEFGHJKMNPQRSTVWXYZ".toCharArray();
    private static final SecureRandom RANDOM = new SecureRandom();

    private IdempotencyKey() {
    }

    public static String generate() {
        return generate(System.currentTimeMillis());
    }

    static String generate(long nowMillis) {
        StringBuilder builder = new StringBuilder(26);
        builder.append(encodeBase32(Math.max(nowMillis, 0L), 10));
        for (int i = 0; i < 16; i++) {
            builder.append(CROCKFORD_BASE32[RANDOM.nextInt(32)]);
        }
        return builder.toString();
    }

    private static String encodeBase32(long value, int length) {
        char[] buffer = new char[length];
        long remaining = value;
        for (int i = length - 1; i >= 0; i--) {
            buffer[i] = CROCKFORD_BASE32[(int) (remaining % 32)];
            remaining /= 32;
        }
        return new String(buffer);
    }
}

File: packages/sdk-kotlin/src/main/kotlin/com/ubag/sdk/IdempotencyKey.kt
package com.ubag.sdk

import java.security.SecureRandom

/** Generates ULID-style idempotency keys: 26 Crockford base32 characters. */
object IdempotencyKey {
    private const val CROCKFORD_BASE32 = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
    private val random = SecureRandom()

    fun generate(nowMillis: Long = System.currentTimeMillis()): String {
        val builder = StringBuilder(26)
        builder.append(encodeBase32(maxOf(nowMillis, 0L), 10))
        repeat(16) {
            builder.append(CROCKFORD_BASE32[random.nextInt(32)])
        }
        return builder.toString()
    }

    private fun encodeBase32(value: Long, length: Int): String {
        val buffer = CharArray(length) { '0' }
        var remaining = value
        for (i in length - 1 downTo 0) {
            buffer[i] = CROCKFORD_BASE32[(remaining % 32).toInt()]
            remaining /= 32
        }
        return String(buffer)
    }
}
